CASSANDRA-19915 Upgrade for the native protocols spec page

tengu-alt opened the pull request:
This PR enhances the Native Protocols page by replacing the plain text with an interactive, section-based navigation. The implementation leverages [cqlprotodoc](https://github.com/martin-sucha/cqlprotodoc) tool that converts CQL protocol specification files into HTML.

A `.sh` script was added to download `cqlprotodoc,` convert the `.spec` files containing the native protocol specifications into HTML pages, and then merge them into a single page.

After that, [native protocols page](https://cassandra.apache.org/_/native_protocol.html) looks like this:
![image](https://github.com/user-attachments/assets/a3d4c7de-0189-42cb-9c76-9021ab38feee)
The navigation is clickable, so you can move to the desired section.
![image](https://github.com/user-attachments/assets/891e481e-ff62-45f0-a1d2-52ab06956e84)

michaelsembwever commented on site-content/process-native-protocol-specs-in-docker.sh:
```diff
@@ -0,0 +1,41 @@
+# Variables
+GO_VERSION="1.20.6"
+GO_TAR="go${GO_VERSION}.linux-amd64.tar.gz"
+GITHUB_REPO_PARSER="https://github.com/martin-sucha/cqlprotodoc.git"  # Replace with the actual parser repo URL
```
I'm hesitant to leave this pointing to a living branch.  Pinning it to a SHA would be safest.

And am curious if we could upstream it to here… (but this can happen separately…)

@martin-sucha, would you be willing to upstream it (to either cassandra-website or cassandra-builds) ?

martin-sucha replied:
Yes, it is better to copy it under `github.com/apache` (I'll leave up to you to which repo). I'm willing to donate the cqlprotodoc code to the Apache software foundation and change the licence to Apache license, Version 2.0.

michaelsembwever replied:
Thanks @martin-sucha.  That's all the approval/consent we need.

@tengu-alt , are you willing (and got the time) to include the contents of the repo in the folder `cqlprotodoc` and adjust accordingly, in this PR ? (don't include its LICENSE file, add to its readme a sentence about where it comes from)

tengu-alt replied:
> Thanks @martin-sucha. That's all the approval/consent we need.
> 
> @tengu-alt , are you willing (and got the time) to include the contents of the repo in the folder `cqlprotodoc` and adjust accordingly, in this PR ? (don't include its LICENSE file, add to its readme a sentence about where it comes from)

Sure! I will do it.

michaelsembwever commented:
This is a solid piece of work, and valuable addition to the website !! 

@ossarga , would you be interested in glancing over this in review, as it touches on your past work ?

smiklosovic commented on cqlprotodoc/README.md:
```diff
@@ -0,0 +1,12 @@
+1. Update template-notice.txt based on your Cassandra distribution.
+2. `cqlprotodoc <path_to_cassandra_doc_dir> <output_dir>`
+
+The generated files are published at https://martin-sucha.github.io/cqlprotodoc/
```
@tengu-alt dont we need to remove this?

tengu-alt replied:
Agree, removed.

smiklosovic commented on cqlprotodoc/template-notice.txt:
```diff
@@ -0,0 +1,5 @@
+Apache Cassandra
+Copyright 2009-2022 The Apache Software Foundation
```
@tengu-alt not sure about the year ... @michaelsembwever thoughts? Do we even need the year at all?

martin-sucha replied:
Do we even need the `template-notice.txt` file anymore? It was present in `cqlprotodoc` because the output of conversion of the Apache2-licensed text was published standalone on a separate website.

The Native Protocols page on apache.org already has ```© 2009-2024 The Apache Software Foundation under the terms of the Apache License 2.0. Apache, the Apache feather logo, Apache Cassandra, Cassandra, and the Cassandra logo, are either registered trademarks or trademarks of The Apache Software Foundation.``` in the footer, so I think we could remove the notice altogether.

tengu-alt replied:
Agree, done. 
Also, I have made some refactoring in `main.go`

michaelsembwever reviewed (approved):
LGTM. Thanks again @tengu-alt !  final approving depends on the other reviewers.

martin-sucha commented on site-content/process-native-protocol-specs-in-docker.sh:
```diff
@@ -0,0 +1,36 @@
+set -e
+
+# Variables
+GO_VERSION="1.20.6"
```
Go 1.20.6 is no longer supported. Please bump to the latest version — 1.23.1.

Going forward, it would be good to set it up in a way so that dependabot or renovatebot could open pull requests to upgrade the version (I don't know if one of those tools is used in this repository though)

tengu-alt replied:
Thank you, fixed.

martin-sucha commented on site-content/source/modules/ROOT/pages/native_protocol.adoc:
```diff
@@ -3,21 +3,81 @@
+
+[source, js]
+++++
+<script>
```
Since we moved the cqlprotodoc code here, I wonder if including the content using Javascript is still needed or whether we should modify the `template.gohtml` / cqlprotodoc output and include the html files directly.

tengu-alt replied:
I think it's better to keep things as they are because the script placed in the .adoc template appears more readable to me. Also, we avoid changing the website template by keeping native_protocol.html as the location where the generated pages are merged.

tengu-alt commented:
> LGTM. Thanks again @tengu-alt ! final approving depends on the other reviewers.

Thanks a lot! Will wait for reviews :)

martin-sucha reviewed (approved):
Thanks! Looks good to me. I'll leave up to you whether to change to directly include the contents instead of using Javascript.

tengu-alt commented:
> Thanks! Looks good to me. I'll leave up to you whether to change to directly include the contents instead of using Javascript.

Thank you!

smiklosovic commented:
going to merge this soon.

smiklosovic commented:
@tengu-alt can  you please tell me how you succeeded to render this locally? All I am getting is this:

![image](https://github.com/user-attachments/assets/51ae1490-06d8-441e-bbad-b9e60214651d)

I checked out your branch and did  `./run.sh website build` and opened this page in the browser:

file:///home/fermat/dev/cassandra/cassandra-website/site-content/build/html/_/native_protocol.html

smiklosovic commented:
I have also run `./run.sh website container` to build the image (not sure why that command is called "container" though) which adds new shell script but the result is the same.

tengu-alt commented:
@smiklosovic you need to build an image from a `Dockerfile` inside of the `site-content` directory and provide the tag:
`docker build . -t <tag>`
And use this image when you run the `./run.sh` script 
`./run.sh website build -c <tag>`

tengu-alt commented:
> I have also run `./run.sh website container` to build the image (not sure why that command is called "container" though) which adds new shell script but the result is the same.

@smiklosovic Has my advice helped solve the build issue?

smiklosovic commented:
@tengu-alt I did what you suggested, it is still same. After your steps I just do 

````
$ open content/_/native_protocol.html
````

firefox opens that patch and I just have same results as shown in the comment from (1)

(1) https://github.com/apache/cassandra-website/pull/282#issuecomment-2345501337

smiklosovic commented:
@michaelsembwever does this render fine for you based on the steps @tengu-alt gave? I am a little bit out of options here.

michaelsembwever commented:
Because of the site-ui/ change in the PR, the way to build this should be
```
./run.sh website container
./run.sh website build
```

This is what's documented in this repo's README, and it is how ci-cassandra.apache.org builds the website (so it needs to work this way). 
ref: https://github.com/apache/cassandra-builds/blob/trunk/jenkins-dsl/cassandra_job_dsl_seed.groovy#L1401-L1414

This _looks_ to work for me, at least i'm getting the `content/_/native_protocol_v*.html` pages.
But the `content/_/native_protocol.html` is still the same as Stefan's screenshot above.
build logs in https://gist.github.com/michaelsembwever/ae8611f4d62e227d02aef9fc89e64657

(It would also be nice to turn off the progress reporting as go downloads…)

tengu-alt commented:
@smiklosovic Now I see. There are some issues with `iFrame` and document origin. I will fix it.

tengu-alt commented:
@smiklosovic I've done some research, and it turns out that the browser does not show the content because the page is loaded using the `file://` protocol (when you open it via `open`, `google-chrome` etc.) which is treated as `null` origin by the browser, and content is blocked by the `CORS` policy .
When I open the file by IDE(which serve it over `localhost:`) it's work just fine.
Could you please try running it on the local HTTP server (e.g. open file thought IDE) and let me know if it resolves the issue?

michaelsembwever commented on site-content/source/modules/ROOT/pages/native_protocol.adoc:
```diff
@@ -3,21 +3,81 @@
+[source, js]
+++++
+<div id="contentv3" class="doc-container"></div>
+<iframe id="iframev3" src="native_protocol_v3.html" style="display:none;"></iframe>
```
is there a way to include the generated native_protocol_v3.html page, instead of using an iframe ?

tengu-alt replied:
It can be done using the `fetch` but it's relying on the same `CORS` policy.

michaelsembwever replied:
can it not be included (somehow) at antora build time ?  if possible that would be the best IMO.

tengu-alt replied:
Could you please clarify - did you mean the JavaScript contained within the .adoc file?

michaelsembwever replied:
no. javascript and iframes are all doing the including at view time.

we should be embedding the html contents into the other file at antora build time.

michaelsembwever replied:
this is what i mean: https://github.com/apache/cassandra-website/commit/c7b6b60eb51c4c27e070adb9f3596e979491eb2f 

note, some small things are still needed on top of that^
 (e.g. there's links in the headers of each page that don't work, and we need gitignore if we have to put the generated files under site-content)
 
(feel free to squash that commit into your work here)

tengu-alt replied:
Understood. I'll try to move it into the `.hbs` layout.

michaelsembwever replied:
Also… there's another thing we might want cleaned up here.

The native protocol spec files are actually here:
https://github.com/apache/cassandra/tree/trunk/doc

They're the source of truth, and those that will be kept up to date.

IUUC the ones in cassandra-website are unnecessary duplicates: https://github.com/apache/cassandra-website/tree/trunk/site-content/source/modules/ROOT/examples/TEXT

We can assume that these from apache/cassandra:trunk are accurate. And delete the `examples/TEXT` directory.    And then it might be possible then to not need any gitignore addition.

Going further, we could also remove the native_protocol.adoc file from casandra-website altogether, move it to here instead: https://github.com/apache/cassandra/tree/trunk/doc/modules/cassandra/pages 

(this would mean running `process-native-protocol-specs-in-docker.sh` on each `generate_cassandra_versioned_docs()`)

tengu-alt replied:
Sounds great! I'll apply it.

tengu-alt commented:
@smiklosovic I have to update my solution: for some reason the files that we need for the  complete page render are not stored under the `site-content` directory but under the `content`(I'm gonna fix it) and you have to look the `native_protocol.html` under the `content` dir. (solution about CORS policy and the `file://` protocol is actual)

smiklosovic commented:
@tengu-alt I would prefer to wait until the complete / working solution is in place, if you do not mind.

tengu-alt commented:
@michaelsembwever I have moved the HTML and JavaScript from the `native_protocol.adoc` file into the new .hbs layout, which is now responsible for it. Additionally, the native_protocol.spec files are currently fetched from the [Apache Cassandra GitHub repository](https://github.com/apache/cassandra/tree/trunk/doc).

The issue with the headers arises because the `native_protocol_v*.spec` files contain typos that cause problems.
(e.g. https://github.com/apache/cassandra/blob/trunk/doc/native_protocol_v3.spec#L719-L740) 
I can create a pull request for Cassandra to fix these typos.

Could you also clarify why we need to move the `native_protocol.adoc` file into the Cassandra repository and run the `process-native-protocol-specs-in-docker.sh` on each `generate_cassandra_versioned_docs()` execution? As I understand it, the native protocol spec page is identical for all Cassandra versions.

tengu-alt commented:
> @tengu-alt I would prefer to wait until the complete / working solution is in place, if you do not mind.

The `native_protocol_v*.spec` files now appears in both `content` and `site-content` dirs. (solution about `CORS` policy and the `file://` protocol is actual)

michaelsembwever commented:
> I can create a pull request for Cassandra to fix these typos.

Please do, thanks!

> Could you also clarify why we need to move the native_protocol.adoc file into the Cassandra repository and run the process-native-protocol-specs-in-docker.sh on each generate_cassandra_versioned_docs() execution?

The .spec files are originally found under https://github.com/apache/cassandra/tree/trunk/doc/
They won't be necessarily be identical on each branch.  i.e. when `v6` is introduced it will only be found initially in the `trunk` branch.

tengu-alt commented:
> > I can create a pull request for Cassandra to fix these typos.
> 
> Please do, thanks!

Done: https://github.com/apache/cassandra/pull/3714

tengu-alt commented:
> > I can create a pull request for Cassandra to fix these typos.
> 
> Please do, thanks!
> 
> > Could you also clarify why we need to move the native_protocol.adoc file into the Cassandra repository and run the process-native-protocol-specs-in-docker.sh on each generate_cassandra_versioned_docs() execution?
> 
> The .spec files are originally found under https://github.com/apache/cassandra/tree/trunk/doc/ They won't be necessarily be identical on each branch. i.e. when `v6` is introduced it will only be found initially in the `trunk` branch.

Understood, working on it.

michaelsembwever commented:
> The .spec files are originally found under https://github.com/apache/cassandra/tree/trunk/doc/ They won't be necessarily be identical on each branch. i.e. when v6 is introduced it will only be found initially in the trunk branch.

I've taken your latest work @tengu-alt and working it a bit.
Expect the patch-on-patch from me either in Dec' or in the new year.

tengu-alt commented:
> > The .spec files are originally found under https://github.com/apache/cassandra/tree/trunk/doc/ They won't be necessarily be identical on each branch. i.e. when v6 is introduced it will only be found initially in the trunk branch.
> 
> I've taken your latest work @tengu-alt and working it a bit. Expect the patch-on-patch from me either in Dec' or in the new year.

I am currently Implementing the protocol pages generation on each version(Maybe it will not be done exactly inside the `generate_cassandra_versioned_docs` ). 
Thanks for the help! What kind of patch do I need to expect?

michaelsembwever committed:
SQUASH – wip – generate spec html pages for each intree version

michaelsembwever commented:
> What kind of patch do I need to expect? 

Ah, i misunderstood you – didn't realise you were still in-progress… 
this is the work-in-progress
https://github.com/tengu-alt/cassandra-website/commit/c06aa65f05b71f40213f69e4444c3ff57703c528 
it's still a bit messy, but maybe you get the drift.

tengu-alt commented:
> > What kind of patch do I need to expect?
> 
> Ah, i misunderstood you – didn't realise you were still in-progress… this is the work-in-progress [tengu-alt@c06aa65](https://github.com/tengu-alt/cassandra-website/commit/c06aa65f05b71f40213f69e4444c3ff57703c528) it's still a bit messy, but maybe you get the drift.

The `process-native-protocol-specs-in-docker.sh` will look differently, and will be called not from the `generate_cassandra_versioned_docs` because the script needs to be run after the `render_site_content_to_html`, which cleans all generated files.
It takes the `.spec` files from `Cassandra/doc`, and `site-content/build/html/Cassandra/$DOC_VERSION/cassandra/native-protocol` is the output dir(I've created the `native-protocol` dir in the Cassandra repo, so we can have different native protocol doc pages for the different Cassandra versions 
Let's apply your changes from `work-in-progress` (e.g. Dockerfile changes) after my next patch if you don't mind.

michaelsembwever commented:
> Let's apply your changes from work-in-progress (e.g. Dockerfile changes) after my next patch if you don't mind. 

Go for it.  It's only for reference sake, and might not be used at all depending on what you land 👍

tengu-alt commented:
@michaelsembwever I have done some refactoring on the `process-native-protocol-specs-in-docker.sh`.
From now, the script is called only if `generate docs` command for build is running (also implemented it for preview mode, but it throws some errors when `generate docs` is running. I've also tested it on the master branch, and it's the same).
I also moved the `native_protocol.adoc` file to the Cassandra Doc under the `native-protocol` directory, and antora generates a page using the `.hbs` layout from the website. The `native_protocol.adoc` needs to be in the Cassandra repo.

The `generate_native_protocol_specs_pages()` now iterates over specified Cassandra versions and calls the script if the provided version of Cassandra contains the `native-protocol` dir under the `doc`. 
After that generated pages are put into the same versioned doc directory under the `site-content` and copied into the `content`.

The `generate_native_protocol_specs_pages()` can't be called with the`generate_cassandra_versioned_docs()` because the script needs to be run after the `render_site_content_to_html` because antora removes the generated files.

(I also added the license part inside script)

tengu-alt committed:
Upgrade for the native protocols spec page

tengu-alt commented:
Added `.hbs` layout enhancement. From now, it is more easy to add a new spec: just need to add a file name into the array.
The `iFrame` creation is on `js`.

michaelsembwever commented:
Looking good @tengu-alt .

I've taken the liberty to add the following via https://github.com/thelastpickle/cassandra-website/tree/mck/native-protocols-page-upgrade and https://github.com/thelastpickle/cassandra/tree/mck/native-protocol-spec-fix :
 - rebased off trunk (so I can get it working on arm64 )
 - add jdk17 support (also adds C* 5.0 jdk auto-detect)
 - generate html files into assets/attachments directory
 - move `process-native-protocol-specs-in-docker.sh` to be a part of `ant gen-asciidoc`
   - by first moving `process-native-protocol-specs-in-docker.sh` in-tree 
   - and building cqlprotodoc off a shallow clone of cassandra-website – as we don't want to introduce golang in-tree just yet
 - add doc/modules/cassandra/pages/reference/native-protocol.adoc in-tree

This generates the per-version html files under attachments, e.g. the url `doc/5.1/cassandra/_attachments/native_protocol_v5.html`

And the aggregated native-protocol.html as `doc/5.1/cassandra/reference/native-protocol.html`

Otherwise, I'm not seeing `native-protocol.hbs` being taken advantage of yet.  (Do we even need it?)

Screenshot of native-protocol.html
<img width="1355" alt="Screenshot 2024-12-30 at 15 53 42" src="https://github.com/user-attachments/assets/2c0ae6de-f706-4004-b840-179529dcd2e4" />

michaelsembwever commented:
> I've taken the liberty to add the following via…

On my (copy of your) cassandra branch i've added another comment to autogenerate `modules/cassandra/pages/reference/native-protocol.adoc`

If we agree to put this under reference, the nav bar needs to be fixed/cleaned too…

tengu-alt commented:
@michaelsembwever Thank you for your work! 
I have looked at and tested your approach(including the `native-protocol.adoc` autogeneration), and It looks like something will be cool to apply! (I also liked how the tabulation in the `table of contents` looks). 

Regarding your question about the `.hbs` layout: it manages the logic for inserting versioned page content and navigating within sections using the `table of contents`. This is necessary because, without it, a link to a paragraph within a section might incorrectly navigate to the same paragraph in a previous section. This issue appears on your solution, and when i tried to update the ui-bundle with `.hbs` layout it started to conflict with your approach.

If `native-protocol.adoc` is going to be autogenerated, we can just put the navigation logic (after some refactoring, because at this moment it depends on the `iFrame` logic) within the `process-native-protocol-specs-in-docker.sh`. It seems to work fine without `iFrames` and `.hbs` layout as well (I only noticed the navigation issue), so, i think, we can get rid of it.

>If we agree to put this under reference, the nav bar needs to be fixed/cleaned too…

Which one nav bar do you mean? I don't clearly understand.

michaelsembwever commented:
if i understand you correctly, then i make the suggestion:
can we simplify all this by keeping the versioned pages separate, and making the autogenerated `native-protocol.adoc` page just link to them rather than include them all in one ? 

This allows us to remove the `.hbs` layout (and any use of iframe).

Or do you want to do the refactoring to `process-native-protocol-specs-in-docker.sh` so it fixes the nav links (and doesn't need iframes) ? 


> Which one nav bar do you mean? I don't clearly understand. 

Separate issue in the left-side main navigation column.  See this screenshot, the list don't match, and what's in the left-side just doesn't make sense.

![Screenshot 2025-01-08 at 07 32 38](https://github.com/user-attachments/assets/9785991e-a439-4ab2-b72d-0ef9eeb6f468)

tengu-alt commented:
> Or do you want to do the refactoring to process-native-protocol-specs-in-docker.sh so it fixes the nav links (and doesn't need iframes) ?

This one. 
I meant to add the onload js function after the part where you add the versioned files from the attachment into the summary file. 
Also, we need to clear the specs from unnecessary headers like this one below: 
![image](https://github.com/user-attachments/assets/1f9738ab-fbd3-4ec2-8e27-1f20d2945ae1)
And those artifacts: 
![image](https://github.com/user-attachments/assets/8374074f-d1f2-4f5f-92e2-afcaa175e517)
Clearing those things also was the part of the .hbs.

Regarding the separate issue: I'll try to figure out what can cause this problem. Does it need to be part of this PR?

michaelsembwever commented:
> This one. 

👍


> Regarding the separate issue: I'll try to figure out what can cause this problem. Does it need to be part of this PR? 

No.

tengu-alt commented:
> 👍 

Understood. So how is it going to be? Do I need to expect your patch here and here: https://github.com/apache/cassandra/pull/3714, and after I'll add the enhancements for the `.sh`script?
Or how is this usually done?

michaelsembwever commented:
> So how is it going to be? Do I need to expect your patch here and here: https://github.com/apache/cassandra/pull/3714, and after I'll add the enhancements for the .sh script?

Yeah.

1. Merge (cherry-pick) https://github.com/tengu-alt/cassandra-website/commit/dcae2d8fb411c626c62271dc38fcb23920fc912c into this PR 

and

2. merge (cherry-pick) https://github.com/apache/cassandra/commit/a4ffdba561a6bc31ba8cd540445d35cf92272344 and https://github.com/apache/cassandra/commit/9cadbe197ad502ac93daf60326f727e5eb73270d into your https://github.com/apache/cassandra/pull/3714

Both (1) and (2) are expected to be squashed into your original commits.  You can do this now or towards the end of the review cycle.

(It is just your two PRs that will eventually be merged into the project.)

tengu-alt commented:
> > So how is it going to be? Do I need to expect your patch here and here: [apache/cassandra#3714](https://github.com/apache/cassandra/pull/3714), and after I'll add the enhancements for the .sh script?
> 
> Yeah.
> 
> 1. Merge (cherry-pick) [tengu-alt@dcae2d8](https://github.com/tengu-alt/cassandra-website/commit/dcae2d8fb411c626c62271dc38fcb23920fc912c) into this PR
> 
> and
> 
> 2. merge (cherry-pick) [apache/cassandra@a4ffdba](https://github.com/apache/cassandra/commit/a4ffdba561a6bc31ba8cd540445d35cf92272344) and [apache/cassandra@9cadbe1](https://github.com/apache/cassandra/commit/9cadbe197ad502ac93daf60326f727e5eb73270d) into your [CASSANDRA-20117 Native protocol spec fix cassandra#3714](https://github.com/apache/cassandra/pull/3714)
> 
> Both (1) and (2) are expected to be squashed into your original commits. You can do this now or towards the end of the review cycle.
> 
> (It is just your two PRs that will eventually be merged into the project.)

Got it! Thank you.

tengu-alt committed:
SQUASH - remove .hbs layout

tengu-alt commented:
@michaelsembwever The `.hbs` layout is removed. I've updated the https://github.com/apache/cassandra/pull/3714 with navigation logic within `.sh` script and fixed the link to the page from documentation `index.adoc`(the page is unresolved when it is not generated yet, and asciidoctor alters link when we use `xref:` so i switched i with `link:` and it works )

tengu-alt commented:
> Separate issue in the left-side main navigation column. 

I have created PR into the Cassandra repo regarding the issue with nav bar
https://github.com/apache/cassandra/pull/3787

michaelsembwever commented:
> The .hbs layout is removed. I've updated the https://github.com/apache/cassandra/pull/3714 with navigation logic within .sh script and fixed the link to the page from documentation index.adoc(the page is unresolved when it is not generated yet, and asciidoctor alters link when we use xref: so i switched i with link: and it works ) 

I lost track @tengu-alt .  Is this just waiting on me now ?

tengu-alt commented:
@michaelsembwever I have updated PRs with your commits and added my updates on top of that. As I understand it awaits further review. 
After that commits here and [here](https://github.com/apache/cassandra/pull/3714) are going to be squashed and merged. Is it correct?

michaelsembwever reviewed (approved):
LGTM.

I had to add this commit https://github.com/apache/cassandra/commit/f4128e0cbfdfa6f908da3b058d3a23d8aba77360 to get it to work, but it's trivial and I can squash that in when I commit (the fixme need line for the git clone needs adjusting too)

michaelsembwever commented:
I'll commit this tomorrow @tengu-alt, squashing as I do it (keeping you as the author ofc).  Thanks for your patience and persistence on both branches!

tengu-alt commented:
> I'll commit this tomorrow @tengu-alt, squashing as I do it (keeping you as the author ofc). Thanks for your patience and persistence on both branches!

Great! Thanks a lot for your work!

michaelsembwever commented:
committed as https://github.com/apache/cassandra-website/commit/2fefee9b5135faf69ea9634de128c39871979c58
